src/actions/FirstOpenActions.tsx

```diff
@@ -1,25 +1,46 @@
 import { makeReactNativeDisklet } from 'disklet'
 
 import { FIRST_OPEN } from '../constants/constantSettings'
+import { makeUuid } from '../util/utils'
 
 const firstOpenDisklet = makeReactNativeDisklet()
 
-let isFirstOpen: 'true' | 'false'
+let isFirstOpen: 'true' | 'false' | undefined
+let deviceId: string | undefined
+let firstOpenEpoch: number | undefined
 
 /**
  * Returns whether this session was the first time the user opened the app.
- * Repeated calls will return the same result.
+ * Repeated calls will return the same result. Also sets the deviceId
  */
-export const getIsFirstOpen = async () => {
-  if (isFirstOpen != null) return isFirstOpen
+export const getFirstOpenInfo = async () => {
+  // Already have the values
+  if (isFirstOpen != null && deviceId != null && firstOpenEpoch != null) return { isFirstOpen, deviceId, firstOpenEpoch }
   else {
+    // Generate new values
+    let firstOpenText
     try {
-      await firstOpenDisklet.getText(FIRST_OPEN)
+      firstOpenText = await firstOpenDisklet.getText(FIRST_OPEN)
       isFirstOpen = 'false'
     } catch (error: any) {
       await firstOpenDisklet.setText(FIRST_OPEN, '')
       isFirstOpen = 'true'
     }
-    return isFirstOpen
+
+    // Not the first time they opened the app, but with an older version that
+    // didn't set a deviceId and firstOpen date
+    let deviceId
+    let firstOpenEpoch
+    if (firstOpenText == null || firstOpenText === '') {
+      deviceId = makeUuid()
+      const firstOpenEpoch = Math.floor(new Date().getTime())
+      await firstOpenDisklet.setText(FIRST_OPEN, JSON.stringify({ deviceId, firstOpenEpoch }))
+    } else {
+      const firstOpenInfo = JSON.parse(firstOpenText)
```

Comment:
Add cleaner so we don't trust data from disk which may be corrupt. This also gives matching types for your return value